HIP-3.md

```diff
@@ -0,0 +1,223 @@
+        uint32 _destinationDomain,
+        uint256 _gas,
+        address _refundAddress
+    ) external payable;
+}
+```
+
+An IGP that quotes and requires on-chain payment for a total amount of gas that is to be used by the ISM and recipient `handle` would look like:
+
+```
+interface FeeQuotingIgp {
+    /**
+     * @notice Emitted when a gas payment is made.
+     * @param messageId The message ID.
+     * @param amount The amount of native tokens paid.
+     * @param gas The amount of destination gas paid for.
+     */
+    event GasPayment(
+        bytes32 indexed messageId,
+        uint256 amount,
+        uint256 gas
+    );
+
+    /**
+     * @notice Accepts payment for a message, reverting if not enough payment was provided.
+     * @dev The `_ismAndHandleGas` should cover ISM metadata calldata gas, ISM.verify gas,
+     *   and gas required by the recipient's `handle` function.
+     *   The gas required for other operations involving the processing of the
+     *   message is read from storage and is added to `_ismAndHandleGas` to calculate
+     *   the total estimated amount of destination gas that's required.
+     *   The stored value includes intrinsic gas, approximated message calldata
+     *   gas, and any overhead gas incurred by the Mailbox contract.
+     *   Using the total expected destination gas usage, the destination gas price
+     *   (from an oracle), and the destination chain native token quoted in the origin
+     *   chain native token (from an oracle), the required fee payment is calculated
+     *   and required to be at least the msg.value. Any overpayment is sent back
+     *   to the `_refundAddress`.
+     * @dev Emits the `GasPayment` event with the total amount of gas paid for.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _ismAndHandleGas The amount of gas to cover ISM metadata calldata,
+     *   ISM.verify gas, and gas required by the recipient's `handle` function.
+     * @param _refundAddress The address to refund any msg.value overpayment to.
+     */
+    function payForGas(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _ismAndHandleGas,
+        address _refundAddress
+    ) external payable;
+}
+```
+
+Hyperlane users are likely to be confident in the amount of gas that their recipient's `handle` function will use on the destination, but they likely don't have the same confidence in the amount of gas they need for their ISM. Users may be using the default ISM which could change entirely, or their ISM may undergo parameter changes that could change the amount of gas required. To provide a better experience, ISM-specific IGPs can be provided that allow users to supply the amount of gas their recipient's `handle` function consumes, and the ISM-specific costs are read from storage and accurately updated by the IGP owner.
+
+```
+contract IsmFeeQuotingIgp {
```

Comment:
Is the idea that there just be one of these that represents default ISMs everywhere? Or do you expect everyone that deploys an ISM to also deploy and configure one of these?

Reply:
I think we'd want to provide one that represents default ISMs everywhere. And permissionless deployment chains can choose to do the same. And then if someone deploys their own non-default ISM, they can choose to deploy and configure one of these or skip it entirely and go with your suggestion of requiring applications to be aware of the ISM costs themselves. I don't think we need to be prescriptive with it when it comes to non-default ISMs